From 802d686fe7d5d23e6c4ffc95a4b099d52441cb24 Mon Sep 17 00:00:00 2001
From: David Gibson <dgibson@redhat.com>
Date: Mon, 15 Sep 2014 07:13:18 -0400
Subject: [virt] kvm/ppc: book3s - Always use the saved DAR value

Message-id: <1410765214-16377-63-git-send-email-dgibson@redhat.com>
Patchwork-id: 94760
O-Subject: [PATCH 62/78] KVM: PPC: BOOK3S: Always use the saved DAR value
Bugzilla: 1123145 1123133 1123367
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Don Zickus <dzickus@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>

Signed-off-by: David Gibson <dgibson@redhat.com>

commit 7310f3a5b0ecc7ac8b57e70fe395968cbe0fc40a
Author: Aneesh Kumar K.V <aneesh.kumar@linux.vnet.ibm.com>
Date:   Mon May 12 17:04:05 2014 +0530

    KVM: PPC: BOOK3S: Always use the saved DAR value

    Although it's optional, IBM POWER cpus always had DAR value set on
    alignment interrupt. So don't try to compute these values.

    Signed-off-by: Aneesh Kumar K.V <aneesh.kumar@linux.vnet.ibm.com>
    Signed-off-by: Alexander Graf <agraf@suse.de>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/kvm/book3s_emulate.c b/arch/powerpc/kvm/book3s_emulate.c
index 9bdff15..61f38eb 100644
--- a/arch/powerpc/kvm/book3s_emulate.c
+++ b/arch/powerpc/kvm/book3s_emulate.c
@@ -676,6 +676,12 @@ u32 kvmppc_alignment_dsisr(struct kvm_vcpu *vcpu, unsigned int inst)
 
 ulong kvmppc_alignment_dar(struct kvm_vcpu *vcpu, unsigned int inst)
 {
+#ifdef CONFIG_PPC_BOOK3S_64
+ /*
+  * Linux's fix_alignment() assumes that DAR is valid, so can we
+  */
+ return vcpu->arch.fault_dar;
+#else
  ulong dar = 0;
  ulong ra = get_ra(inst);
  ulong rb = get_rb(inst);
@@ -700,4 +706,5 @@ ulong kvmppc_alignment_dar(struct kvm_vcpu *vcpu, unsigned int inst)
  }
 
  return dar;
+#endif
 }
-- 
1.7.1